F811 false positive in match statement

Hi,

really love ruff, such a great tool! It also inspired me to learn Rust ;)

I found the following bug, which can be reproduced with this MWE:
```python
def no_redef(value):
    if value:

        def fun(x, y):
            return x

    else:

        def fun(x, y):
            return y

    return fun


def redef(value):
    match value:
        case True:

            def fun(x, y):
                return x

        case False:

            def fun(x, y):
                return y

    return fun
```
Running ruff throws the following error which (IMO) is a bug:
```bash
> ruff redef.py --isolated
redef.py:24:17: F811 Redefinition of unused `fun` from line 19
Found 1 error.
```
**Why is this a bug?** 
In my understanding both the `no_redef` and `redef` function have the same behavior. Thus, F811 should behave the same for both functions, which is not the case.


This behavior is consistent with flake8/pyflakes, however, I think this is a bug in their code as well and I opened a ticket there too (https://github.com/PyCQA/pyflakes/issues/771).


**Ruff version**
```bash
> ruff --version
ruff 0.0.261
```

**pyproject.toml**
```toml
[tool.ruff]
select = ["F"]
target-version = "py38"
```

Constantin

--- a/crates/ruff/resources/test/fixtures/pyflakes/F811_22.py
+++ b/crates/ruff/resources/test/fixtures/pyflakes/F811_22.py
@@ -1,0 +1,13 @@
+def redef(value):
+    match value:
+        case True:
+
+            def fun(x, y):
+                return x
+
+        case False:
+
+            def fun(x, y):
+                return y
+
+    return fun

--- a/crates/ruff/src/rules/pyflakes/mod.rs
+++ b/crates/ruff/src/rules/pyflakes/mod.rs
@@ -96,6 +96,7 @@ mod tests {
     #[test_case(Rule::RedefinedWhileUnused, Path::new("F811_19.py"); "F811_19")]
     #[test_case(Rule::RedefinedWhileUnused, Path::new("F811_20.py"); "F811_20")]
     #[test_case(Rule::RedefinedWhileUnused, Path::new("F811_21.py"); "F811_21")]
+    #[test_case(Rule::RedefinedWhileUnused, Path::new("F811_22.py"); "F811_22")]
     #[test_case(Rule::UndefinedName, Path::new("F821_0.py"); "F821_0")]
     #[test_case(Rule::UndefinedName, Path::new("F821_1.py"); "F821_1")]
     #[test_case(Rule::UndefinedName, Path::new("F821_2.py"); "F821_2")]

--- a/crates/ruff_python_ast/src/branch_detection.rs
+++ b/crates/ruff_python_ast/src/branch_detection.rs
@@ -71,6 +71,10 @@ fn alternatives(stmt: RefEquality<Stmt>) -> Vec<Vec<RefEquality<Stmt>>> {
                 body.iter().map(RefEquality).collect()
             }))
             .collect(),
+        StmtKind::Match { cases, .. } => cases
+            .iter()
+            .map(|case| case.body.iter().map(RefEquality).collect())
+            .collect(),
         _ => vec![],
     }
 }

